pravega/pravega
# Issue 2641: Stream re-creation fix

**Change log description**
This PR allows users to re-create streams: 
- The stream entity in the controller is now able to create segment numbers from a given `startingSegmentNumber` onward. This will avoid naming collisions with a previously deleted stream with the same name whose segments are being asynchronously deleted from the segment store. 
- The `AbstractStreamMetadataStore.java` implementations (`ZKMetadataStore.java`, `InMemoryMetadataStore.java`) are responsible for i) persistently storing the highest segment number for a stream (upon a `deleteStream` operation), and ii) to get a safe starting number (upon a `createStream` operation).
- `StreamMetadataTasks.java` takes into account the safe starting segment number calculated by `AbstractStreamMetadataStore.java` to create new segments in the segment store.

**Purpose of the change**
Fixes #1012, #1681, and #2641.

**What the code does**
Until now, all the streams were assumed to start the ids of their segments by `0`; this is a simple assumption but it cannot handle stream re-creation within short time intervals, given that segments in the segment store are deleted asynchronously. That is, if we delete a stream and create another with the same name, the controller could inadvertently ask the segment store to create a segment that is still _being deleted_ and related to the previously deleted stream. To fully support stream re-creations, we need the system to pick a starting segment number for a stream that is higher than the highest segment of the previously deleted stream with the same name. To achieve this, this PR addresses the following points:

- Stream objects in the controller now require a new metadata item: `startingSegmentNumber`. This number will allow the controller to start creating segment numbers for a given stream from `startingSegmentNumber` onward to avoid naming collisions with a previously deleted stream with the same name whose segments are being asynchronously deleted from the segment store. Consequently, `Stream.java` API (controller) adds a new method (`getStartingSegmentNumber`) and modifies the `create` method by adding a new argument (`startingSegmentNumber`).

- `PersistentStreamBase` implicitly benefits `startingSegmentNumber` to get the appropriate segment indexes. That is, for a `startingSegmentNumber > 0` the system creates `startingSegmentNumber` zero entries in the `segmentIndex`. This allows us to avoid modifying most of the logic in `TableHelper`, `SegmentIndexRecord` and `PersistentStreamBase`, as well as to keep the metadata format backward compatible. `PersistentStreamBase` also exposes a new method `getStartingSegmentNumber()` that can be used to get the starting segment number for a stream.

- `AbstractStreamMetadataStore.java` defines two new abstract methods: `recordLastStreamSegment` and `getSafeStartingSegmentNumberFor`. Its respective implementations (`ZKMetadataStore.java`, `InMemoryMetadataStore.java`) are responsible for implementing these methods that are intended to i) persistently store the highest segment number for a stream (upon a `deleteStream` operation) and ii) to get a safe starting number (upon a `createStream` operation). In ZooKeeper, the highest segment number for a stream is stored under `/lastActiveStreamSegment/{stream_path}/{number}` and it remains after the stream has been deleted. Note that this PR does not garbage collects this metadata; this might be achieved with the garbage collection mechanism proposed in PR #2756. 

- While orchestrating the creation workflow for a stream, `StreamMetadataTasks.java` takes into account the result from `getSafeStartingSegmentNumberFor` (encapsulated in a `CreateStreamResponse` object) to safely create new segments in the segment store.

_Parallel stream creations/deletions_: Parallel stream creations and deletions should work in a similar way as they do right now. That is, upon parallel creations, the operation that first hits Zookeeper is the one that will proceed with the creation of the stream, as `PersistentStreamBase` checks this in first place (`checkStreamExists`). All these parallel operations will read the same value from `getSafeStartingSegmentNumberFor`. Once the stream is created, parallel deletions will attempt to delete the Zookeeper metadata tree. Moreover, these parallel deletions may attempt to update the `/lastActiveStreamSegment/{stream_path}/{number}` value, which will be updated only if the new input value represents a segment number higher than the existing one. This is important to ensure that the stream re-creation has a starting segment number `>` than the highest segment of the previously deleted stream. In the case of simultaneous `create` and `delete` operations, we can assume that the first operation to be served is `create` (otherwise `delete` will not even start). Once the `create` operation starts, it sets the metadata of the stream (i.e., `segmentIndex`, `historyTable`) for the new segments to be created. Once the metadata set up is finished, the controller asynchronously creates the segments in the segment store. Let's assume that while the controller is creating segments in the segment store, there is a `delete` operation. In this case, the number of created segments in the segment store will always be lower than the number of initial segments defined in the controller metadata. As we use the controller metadata to get the highest stream segment number upon a `delete` (`getActiveSegments()` call), this means that the `delete` operation will store a `/lastActiveStreamSegment/{stream_path}/{number} >` the highest segment created in the segment store, which will allow us to re-create the stream.

**How to verify it**
All the tests working on new streams (non-recreated) should work as before (unit/integration/system).
Modified unit tests in `TableHelperTests.java` to check the operation of this class with `startingSegmentNumbers>0`.
Added new unit tests in `StreamMetadataStoreTest.java` to check the new operations in `StreamMetadataStore`.
Created new integration test `StreamRecreationTest.java` to check successive re-creations on the same stream.
Modified integration test `EndToEndTruncationTest.java` to test truncation on re-created streams.
Modified integration test `EndToEndWithScaleTest.java` to test scale events on re-created streams.
Updated system test `StreamsAndScopesManagementTest.java` to operate on re-created streams.
